README.md
# Midjourney Archive Chrome Extension

I got a bit tired of Midjourney's limited and buggy archive download solution so
I decided to put together a quick and dirty chrome extension with GPT-4.

Make sure you're logged into Midjourney, otherwise this won't work.

For feedback, reach out to me on [Twitter](https://twitter.com/dreamingtulpa).

## Metadata

Metadata for each job can be found in the `archived_jobs.json` file that gets
added to each generated zip file. To associate jobs to the image files, I've
added an `_archvied_files` attribute to each job that contains the downloaded image file names.

[@GanWeaving](https://twitter.com/ganweaving) put together a
[python script](https://github.com/GanWeaving/inject-midjourney-command) that converts the images to JPEG and injects the metadata into the EXIF data.

## TODO

[ ] Add image settings (prompt, aspect ratio etc) as metadata/properties directly to the png images.

## Development

After modifying `src/popup.js`, run `npm run build` to update `dist/popup.js`.


CHANGELOG.md
# 1.3

- Added an archived_jobs.json file to the zip file that contains all job data

# 1.2

- Updated filename format {YYYY-MM-DD-HHMMSS}_{id}_{truncated_prompt}
- Added option to download all images from v5 grids instead of only upscales

# 1.1

- Moved from popup to tab form

# 1.0

- Initial upscale archive downloader


src/popup_new.js
import JSZip from "jszip";

document.getElementById("dateForm").addEventListener("submit", async (event) => {
  event.preventDefault();

  const startDate = new Date(event.target.startDate.value);
  const endDate = new Date(event.target.endDate.value);
  const upscaleSelection = document.querySelector('input[name="upscaleOptions"]:checked').value;

  // Show progress container and hide form
  const progressContainer = document.getElementById("progressContainer");
  const progressDaysBar = document.getElementById("progressDaysBar");
  const progressDaysMessage = document.getElementById("progressDaysMessage");
  const progressJobsBar = document.getElementById("progressJobsBar");
  const progressJobsMessage = document.getElementById("progressJobsMessage");
  const form = document.getElementById("dateForm");
  form.classList.add("hidden");
  progressContainer.classList.remove("hidden");

  let processedDays = 0;
  const totalDays = Math.floor((endDate - startDate) / (1000 * 60 * 60 * 24)) + 1;

  // Iterate through the dates and perform API calls
  for (let currentDate = startDate; currentDate <= endDate; currentDate.setDate(currentDate.getDate() + 1)) {
    progressDaysMessage.innerText = `[${processedDays}/${totalDays}] Processing ${currentDate.toISOString().slice(0, 10)}...`;
    const year = currentDate.getFullYear();
    const month = currentDate.getMonth() + 1;
    const day = currentDate.getDate();

    // Fetch the archive data
    const archiveResponse = await fetch(`https://www.midjourney.com/api/app/archive/day/?day=${day}&month=${month}&year=${year}`);
    const archiveData = await archiveResponse.json();

    // Create a zip file for the current date
    const zip = new JSZip();
    let fileCount = 0;

    let processedJobs = 0;
    let totalJobs = archiveData.length;

    // Process each item in the archive data
    for (const item of archiveData) {
      // const jobId = item.id;
      const jobId = item;
      progressJobsMessage.innerText = `[${processedJobs}/${totalJobs}] Fetching ${jobId}...`;

      // Fetch the job status data
      const jobStatusResponse = await fetch("https://www.midjourney.com/api/app/job-status/", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ jobIds: [jobId] }),
      });
      const jobStatusData = await jobStatusResponse.json();
      console.log(jobStatusData);

      // Check job version
      let versionNumber = parseFloat(jobStatusData._parsed_params.version);
      let isVersion5Plus = !isNaN(versionNumber) && versionNumber >= 5;

      // Set event type
      let eventType = jobStatusData.event.eventType

      // Process images
      if (isVersion5Plus && eventType !== "upscale" && upscaleSelection === "allImagesV5Grids") {
        fileCount += await processImages(jobStatusData, zip);
      } else if (isVersion5Plus && eventType === "upscale" && upscaleSelection === "onlyV5Upscales") {
        fileCount += await processImages(jobStatusData, zip);
      } else if (!isVersion5Plus && eventType === "upscale") {
        fileCount += await processImages(jobStatusData, zip);
      }

      processedJobs++;
      progressJobsBar.value = (processedJobs / totalJobs) * 100;
      progressJobsMessage.innerText = `[${processedJobs}/${totalJobs}] Fetched ${jobId}!`;
    }
    processedDays++;
    progressDaysBar.value = (processedDays / totalDays) * 100;

    // Generate and download the zip file
    if (fileCount > 0) {
      const zipBlob = await zip.generateAsync({ type: "blob" });
      const downloadUrl = URL.createObjectURL(zipBlob);
      const downloadLink = document.createElement("a");
      downloadLink.href = downloadUrl;
      downloadLink.download = `midjourney_archive_${year}-${month}-${day}_[${fileCount}].zip`;
      downloadLink.click();

      URL.revokeObjectURL(downloadUrl);
    }
  }
  progressDaysMessage.innerText = "Download complete!";
});

async function processImages(jobStatusData, zip) {
  // Download and process images
  const { username, image_paths, id, parent_id, enqueue_time, full_command, prompt, event, _parsed_params } = jobStatusData;
  let fileCount = 0;

  for (const [index, imagePath] of image_paths.entries()) {
    // Fetch the image
    const response = await fetch(imagePath);
    const imageBlob = await response.blob();

    // Filename conversions
    const truncated_username = makeFilenameCompatible(username);
    const truncated_prompt = makeFilenameCompatible(prompt, 48);
    const datetime = convertDateTimeFormat(enqueue_time);

    // Add EXIF metadata
    const modifiedBlob = await addExifMetadata(imageBlob, jobStatusData);

    // Build filename
    let filename;
    if (image_paths.length > 1) {
      filename = `${datetime}_${id}_${index}_${truncated_prompt}.png`
    } else {
      filename = `${datetime}_${id}_${truncated_prompt}.png`
    }

    // Add the image to the zip file
    zip.file(filename, modifiedBlob);
    fileCount++;
  }

  return fileCount;
}

function makeFilenameCompatible(str, maxLength) {
  if (typeof str === 'undefined') {
    str = 'None';
  }
  let result = str.replace(/ /g, '_').replace(/[^\w-]/g, '');
  if (maxLength) {
    return result.length > maxLength ? result.substr(0, maxLength) : result;
  } else {
    return result;
  }
}

function convertDateTimeFormat(dateTimeString) {
  const dt = new Date(dateTimeString);
  const year = dt.getFullYear();
  const month = String(dt.getMonth() + 1).padStart(2, '0'); // Months are 0-based in JavaScript
  const day = String(dt.getDate()).padStart(2, '0');
  const hours = String(dt.getHours()).padStart(2, '0');
  const minutes = String(dt.getMinutes()).padStart(2, '0');
  const seconds = String(dt.getSeconds()).padStart(2, '0');

  return `${year}-${month}-${day}-${hours}${minutes}${seconds}`;
}

async function addExifMetadata(imageBlob, jobStatusData) {
  // Implement this function to add the required EXIF metadata to the image
  // You may use a library like "piexifjs" to manipulate the EXIF data
  return imageBlob;
}

src/popup copy.js
import JSzip from "jszip";

let fileCount = 0;

document.getElementById("dateForm").addEventListener("submit", async (event) => {
  event.preventDefault();

  const startDate = new Date(event.target.startDate.value);
  const endDate = new Date(event.target.endDate.value);
  const upscaleSelection = document.querySelector('input[name="upscaleOptions"]:checked').value;

  const progressContainer = document.getElementById("progressContainer");
  const progressDaysBar = document.getElementById("progressDaysBar");
  const progressDaysMessage = document.getElementById("progressDaysMessage");
  const progressJobsBar = document.getElementById("progressJobsBar");
  const progressJobsMessage = document.getElementById("progressJobsMessage");
  const form = document.getElementById("dateForm");
  form.classList.add("hidden");
  progressContainer.classList.remove("hidden");

  let processedDays = 0;
  const totalDays = Math.floor((endDate - startDate) / (1000 * 60 * 60 * 24)) + 1;

  for (let currentDate = startDate; currentDate <= endDate; currentDate.setDate(currentDate.getDate() + 1)) {
    progressDaysMessage.innerText = `[${processedDays}/${totalDays}] Processing ${currentDate.toISOString().slice(0, 10)}...`;
    const year = currentDate.getFullYear();
    const month = currentDate.getMonth() + 1;
    const day = currentDate.getDate();

    const archiveResponse = await fetch(`https://www.midjourney.com/api/app/archive/day/?day=${day}&month=${month}&year=${year}`);
    const archiveData = await archiveResponse.json();

    const zip = new JSzip();
    let csvData = "Filename,Prompt\n"; // Define CSV headers
    //let fileCount = 0;
    let processedJobs = 0;
    let totalJobs = archiveData.length;

    for (const item of archiveData) {
      const jobId = item;
      progressJobsMessage.innerText = `[${processedJobs}/${totalJobs}] Fetching ${jobId}...`;

      const jobStatusResponse = await fetch("https://www.midjourney.com/api/app/job-status/", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ jobIds: [jobId] }),
      });
      const jobStatusData = await jobStatusResponse.json();

      let versionNumber = parseFloat(jobStatusData._parsed_params.version);
      let isVersion5Plus = !isNaN(versionNumber) && versionNumber >= 5;

      let eventType = jobStatusData.event.eventType

      if (isVersion5Plus && eventType === "imagine" && upscaleSelection === "allImagesV5Grids") {
        let result = await processImages(jobStatusData, zip, csvData);
        fileCount += result.fileCount;
        csvData = result.csvData;
      } else if (isVersion5Plus && eventType === "upscale" && upscaleSelection === "onlyV5Upscales") {
        let result = await processImages(jobStatusData, zip, csvData);
        fileCount += result.fileCount;
        csvData = result.csvData;
      } else if (!isVersion5Plus && eventType === "upscale") {
        let result = await processImages(jobStatusData, zip, csvData);
        fileCount += result.fileCount;
        csvData = result.csvData;
      }

      processedJobs++;
      progressJobsBar.value = (processedJobs / totalJobs) * 100;
      progressJobsMessage.innerText = `[${processedJobs}/${totalJobs}] Fetched ${jobId}!`;
    }

    zip.file("image_data.csv", csvData);

    processedDays++;
    progressDaysBar.value = (processedDays /    // Continued from previous section...
    totalDays) * 100;

    // Generate and download the zip file
    if (fileCount > 0) {  
      const zipBlob = await zip.generateAsync({ type: "blob" });
      const downloadUrl = URL.createObjectURL(zipBlob);
      const downloadLink = document.createElement("a");
      downloadLink.href = downloadUrl;
      downloadLink.download = `midjourney_archive_${year}-${month}-${day}_[${fileCount}].zip`;
      downloadLink.click();

      URL.revokeObjectURL(downloadUrl);
    }
  }

  progressDaysMessage.innerText = "Download complete!";
});

async function processImages(jobStatusData, zip, csvData) {
  const { username, image_paths, id, parent_id, enqueue_time, full_command, prompt, event, _parsed_params } = jobStatusData;
  let fileCount = 0;

  if (!image_paths) {
    console.error('image_paths is null or undefined');
    return;
  }

  for (const [index, imagePath] of image_paths.entries()) {
    const response = await fetch(imagePath);
    const imageBlob = await response.blob();

    const truncated_username = makeFilenameCompatible(username);
    const truncated_prompt = makeFilenameCompatible(prompt, 48);
    const datetime = convertDateTimeFormat(enqueue_time);

    const modifiedBlob = await addExifMetadata(imageBlob, jobStatusData);

    let filename;
    if (image_paths.length > 1) {
      filename = `${datetime}_${id}_${index}_${truncated_prompt}.png`
    } else {
      filename = `${datetime}_${id}_${truncated_prompt}.png`
    }

    zip.file(filename, modifiedBlob);
    fileCount++;

    // Append data to CSV
    csvData += `${filename},"${prompt}"\n`;
  }

  return { fileCount: fileCount, csvData: csvData };
}

function makeFilenameCompatible(str, maxLength) {
  if (typeof str === 'undefined') {
    str = 'None';
  }
  let result = str.replace(/ /g, '_').replace(/[^\w-]/g, '');
  if (maxLength) {
    return result.length > maxLength ? result.substr(0, maxLength) : result;
  } else {
    return result;
  }
}

function convertDateTimeFormat(dateTimeString) {
  const dt = new Date(dateTimeString);
  const year = dt.getFullYear();
  const month = String(dt.getMonth() + 1).padStart(2, '0');
  const day = String(dt.getDate()).padStart(2, '0');
  const hours = String(dt.getHours()).padStart(2, '0');
  const minutes = String(dt.getMinutes()).padStart(2, '0');
  const seconds = String(dt.getSeconds()).padStart(2, '0');

  return `${year}-${month}-${day}-${hours}${minutes}${seconds}`;
}

async function addExifMetadata(imageBlob, jobStatusData) {
  // Implement this function to add the required EXIF metadata to the image
  // You may use a library like "piexifjs" to manipulate the EXIF data
  return imageBlob;
}



src/popup.js
import JSZip from "jszip";

document.getElementById("dateForm").addEventListener("submit", async (event) => {
  event.preventDefault();

  const startDate = new Date(event.target.startDate.value);
  const endDate = new Date(event.target.endDate.value);
  const upscaleSelection = document.querySelector('input[name="upscaleOptions"]:checked').value;

  // Show progress container and hide form
  const progressContainer = document.getElementById("progressContainer");
  const progressDaysBar = document.getElementById("progressDaysBar");
  const progressDaysMessage = document.getElementById("progressDaysMessage");
  const progressJobsBar = document.getElementById("progressJobsBar");
  const progressJobsMessage = document.getElementById("progressJobsMessage");
  const form = document.getElementById("dateForm");
  const error = document.getElementById("error");
  form.classList.add("hidden");
  progressContainer.classList.remove("hidden");
  error.classList.add('hidden')

  let processedDays = 0;
  const totalDays = Math.floor((endDate - startDate) / (1000 * 60 * 60 * 24)) + 1;

  // Iterate through the dates and perform API calls
  for (let currentDate = startDate; currentDate <= endDate; currentDate.setDate(currentDate.getDate() + 1)) {
    progressDaysMessage.innerText = `[${processedDays}/${totalDays}] Processing ${currentDate.toISOString().slice(0, 10)}...`;
    const year = currentDate.getFullYear();
    const month = currentDate.getMonth() + 1;
    const day = currentDate.getDate();

    // Fetch the archive data
    const archiveResponse = await fetch(`https://www.midjourney.com/api/app/archive/day/?day=${day}&month=${month}&year=${year}`);
    if (archiveResponse.status === 403) {
      progressContainer.classList.add("hidden");
      form.classList.remove("hidden");
      error.classList.remove("hidden")
      error.innerText = "Received HTTP 403 Forbidden. It seems you're not logged into https://www.midjourney.com.";
      throw new Error(error.innerText);
    }
    const archiveData = await archiveResponse.json();

    // Create a zip file for the current date
    const zip = new JSZip();
    let fileCount = 0;

    let processedJobs = 0;
    let archivedJobs = []
    let totalJobs = archiveData.length;

    // Process each item in the archive data
    for (const item of archiveData) {
      // const jobId = item.id;
      const jobId = item;
      progressJobsMessage.innerText = `[${processedJobs}/${totalJobs}] Fetching ${jobId}...`;

      // Fetch the job status data
      const jobStatusResponse = await fetch("https://www.midjourney.com/api/app/job-status/", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ jobIds: [jobId] }),
      });
      const jobStatusData = await jobStatusResponse.json();

      console.log(jobStatusData);

      // Check job version
      let versionNumber;
      if (jobStatusData._parsed_params) {
        versionNumber = parseFloat(jobStatusData._parsed_params.niji) || parseFloat(jobStatusData._parsed_params.version);
      } else {
        versionNumber = "undefined";
      }
      let isVersion5Plus = !isNaN(versionNumber) && versionNumber >= 5;

      // Set event type
      let isUpscale = jobStatusData.type !== "grid"

      // Process images
      if (isVersion5Plus && !isUpscale && upscaleSelection === "allImagesV5Grids") {
        fileCount += await processImages(jobStatusData, zip, archivedJobs);
      } else if (isVersion5Plus && isUpscale && upscaleSelection === "onlyV5Upscales") {
        fileCount += await processImages(jobStatusData, zip, archivedJobs);
      } else if (!isVersion5Plus && isUpscale) {
        fileCount += await processImages(jobStatusData, zip, archivedJobs);
      }

      processedJobs++
      progressJobsBar.value = (processedJobs / totalJobs) * 100;
      progressJobsMessage.innerText = `[${processedJobs}/${totalJobs}] Fetched ${jobId}!`;
    }
    processedDays++;
    progressDaysBar.value = (processedDays / totalDays) * 100;

    // Generate and download the zip file
    if (fileCount > 0) {
      zip.file("archived_jobs.json", JSON.stringify(archivedJobs));

      const zipBlob = await zip.generateAsync({ type: "blob" });
      const downloadUrl = URL.createObjectURL(zipBlob);
      const downloadLink = document.createElement("a");
      downloadLink.href = downloadUrl;
      downloadLink.download = `midjourney_archive_${year}-${month}-${day}_[${fileCount}].zip`;
      downloadLink.click();

      URL.revokeObjectURL(downloadUrl);
    }
  }
  progressDaysMessage.innerText = "Download complete!";
});

async function processImages(jobStatusData, zip, archivedJobs) {
  // Download and process images
  const { image_paths, id, enqueue_time, prompt } = jobStatusData;
  let fileCount = 0;

  if (image_paths !== null) {
    jobStatusData._archived_files = [];

    for (const [index, imagePath] of image_paths.entries()) {
      // Fetch the image
      const response = await fetch(imagePath);
      const imageBlob = await response.blob();

      // Filename conversions
      const truncated_prompt = makeFilenameCompatible(prompt, 48);
      const datetime = convertDateTimeFormat(enqueue_time);

      // Add EXIF metadata
      const modifiedBlob = await addExifMetadata(imageBlob, jobStatusData);

      // Build filename
      let filename;
      if (image_paths.length > 1) {
        filename = `${datetime}_${id}_${index}_${truncated_prompt}.png`
      } else {
        filename = `${datetime}_${id}_${truncated_prompt}.png`
      }
      jobStatusData._archived_files.push(filename);

      // Add the image to the zip file
      zip.file(filename, modifiedBlob);
      fileCount++;
    }

    archivedJobs.push(jobStatusData);
  }

  return fileCount;
}

function makeFilenameCompatible(str, maxLength) {
  if (typeof str === 'undefined') {
    str = 'None';
  }
  let result = str.replace(/ /g, '_').replace(/[^\w-]/g, '');
  if (maxLength) {
    return result.length > maxLength ? result.substr(0, maxLength) : result;
  } else {
    return result;
  }
}

function convertDateTimeFormat(dateTimeString) {
  const dt = new Date(dateTimeString);
  const year = dt.getFullYear();
  const month = String(dt.getMonth() + 1).padStart(2, '0'); // Months are 0-based in JavaScript
  const day = String(dt.getDate()).padStart(2, '0');
  const hours = String(dt.getHours()).padStart(2, '0');
  const minutes = String(dt.getMinutes()).padStart(2, '0');
  const seconds = String(dt.getSeconds()).padStart(2, '0');

  return `${year}-${month}-${day}-${hours}${minutes}${seconds}`;
}

async function addExifMetadata(imageBlob, jobStatusData) {
  // Implement this function to add the required EXIF metadata to the image
  // You may use a library like "piexifjs" to manipulate the EXIF data
  return imageBlob;
}
